gpjax/kernels/non_euclidean/categorical.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2022 The JaxGaussianProcesses Contributors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+
+from dataclasses import dataclass
+from typing import NamedTuple, Union
+import jax.numpy as jnp
+from jaxtyping import Float, Int
+import tensorflow_probability.substrates.jax as tfp
+
+from gpjax.base import (
+    param_field,
+    static_field,
+)
+from gpjax.kernels.base import AbstractKernel
+
+from gpjax.typing import (
+    Array,
+    ScalarInt,
+)
+
+tfb = tfp.bijectors
+
+CatKernelParams = NamedTuple(
+    "CatKernelParams",
+    [("stddev", Float[Array, "N 1"]), ("cholesky_lower", Float[Array, " N*(N-1)//2"])],
+)
+
+
+@dataclass
+class CatKernel(AbstractKernel):
+    r"""The categorical kernel is defined for a fixed number of values of categorical input.
+
+    It stores a standard dev for each input value (i.e. the diagonal of the gram), and a lower cholesky factor for correlations.
+    It returns the corresponding values from an the gram matrix when called.
+
+    Args:
+        stddev (Float[Array, "N"]): The standard deviation parameters, one for each input space value.
+        cholesky_lower (Float[Array, "N*(N-1)//2 N"]): The parameters for the Cholesky factor of the gram matrix.
+        inspace_vals (list): The values in the input space this CatKernel works for. Stored for order reference, making clear the indices used for each input space value.
+        name (str): The name of the kernel.
+        input_1hot (bool): If True, the kernel expect to be called with a 1-hot encoding of the input space values. If False, it expects the indices of the input space values.
+
+    Raises:
+        ValueError: If the number of diagonal variance parameters does not match the number of input space values.
+    """
+
+    stddev: Float[Array, " N"] = param_field(jnp.ones((2,)), bijector=tfb.Softplus())
+    cholesky_lower: Float[Array, "N N"] = param_field(
+        jnp.eye(2), bijector=tfb.CorrelationCholesky()
+    )
```

Comment:
it would be nice to be able to control the flexibility,  kinda like when you specify the rank of W in the decomp K = W * W^T + kappa

Reply:
How much work would this be @ingmarschuster? If it's simple, then maybe let's add it to this PR. Otherwise, if you feel it's a good idea, then let's open an issue for it.